openshift/filter.go

```diff
@@ -24,68 +24,88 @@ var availableKinds = []string{
 }
 
 type ResourceFilter struct {
-	Kinds []string
-	Name  string
-	Label string
+	Kinds          []string
+	Name           string
+	Label          string
+	ExcludedKinds  []string
+	ExcludedNames  []string
+	ExcludedLabels []string
 }
 
 // NewResourceFilter returns a filter based on kinds and flags.
 // kindArg might be blank, or a list of kinds (e.g. 'pvc,dc') or
 // a kind/name combination (e.g. 'dc/foo').
 // selectorFlag might be blank or a key and a label, e.g. 'name=foo'.
-func NewResourceFilter(kindArg string, selectorFlag string) (*ResourceFilter, error) {
+func NewResourceFilter(kindArg string, selectorFlag string, excludeFlag string) (*ResourceFilter, error) {
 	filter := &ResourceFilter{
 		Kinds: []string{},
 		Name:  "",
 		Label: selectorFlag,
 	}
 
-	if len(kindArg) == 0 {
-		return filter, nil
-	}
+	if len(kindArg) > 0 {
+		kindArg = strings.ToLower(kindArg)
+
+		if strings.Contains(kindArg, "/") {
+			if strings.Contains(kindArg, ",") {
+				return nil, errors.New(
+					"You cannot target more than one resource name",
+				)
+			}
+			nameParts := strings.Split(kindArg, "/")
+			filter.Name = KindMapping[nameParts[0]] + "/" + nameParts[1]
+			return filter, nil
+		}
 
-	kindArg = strings.ToLower(kindArg)
+		targetedKinds := make(map[string]bool)
+		unknownKinds := []string{}
+		kinds := strings.Split(kindArg, ",")
+		for _, kind := range kinds {
+			if _, ok := KindMapping[kind]; !ok {
+				unknownKinds = append(unknownKinds, kind)
+			} else {
+				targetedKinds[KindMapping[kind]] = true
+			}
+		}
 
-	if strings.Contains(kindArg, "/") {
-		if strings.Contains(kindArg, ",") {
-			return nil, errors.New(
-				"You cannot target more than one resource name",
+		if len(unknownKinds) > 0 {
+			return nil, fmt.Errorf(
+				"Unknown resource kinds: %s",
+				strings.Join(unknownKinds, ","),
 			)
 		}
-		nameParts := strings.Split(kindArg, "/")
-		filter.Name = KindMapping[nameParts[0]] + "/" + nameParts[1]
-		return filter, nil
-	}
 
-	targetedKinds := make(map[string]bool)
-	unknownKinds := []string{}
-	kinds := strings.Split(kindArg, ",")
-	for _, kind := range kinds {
-		if _, ok := KindMapping[kind]; !ok {
-			unknownKinds = append(unknownKinds, kind)
-		} else {
-			targetedKinds[KindMapping[kind]] = true
+		for kind := range targetedKinds {
+			filter.Kinds = append(filter.Kinds, kind)
 		}
-	}
 
-	if len(unknownKinds) > 0 {
-		return nil, fmt.Errorf(
-			"Unknown resource kinds: %s",
-			strings.Join(unknownKinds, ","),
-		)
+		sort.Strings(filter.Kinds)
 	}
 
-	for kind := range targetedKinds {
-		filter.Kinds = append(filter.Kinds, kind)
+	if len(excludeFlag) > 0 {
+		unknownKinds := []string{}
+		excludes := strings.Split(excludeFlag, ",")
+		for _, v := range excludes {
```

Comment:
A strings.ToLower should probably be added.

Reply:
Done in 676271966f1136861eb1b35158cfbdf95c48325e.